LMFDB/lmfdb
# Load L-functions from database for HMF and Artin when available

This PR partially addresses #2754 and #2873 by having the L-function code check the instances database when displaying HMF and Artin L-functions.  If the L-function is stored in the database it will be used, otherwise it will revert to computing them on the fly as it does now.  To keep the code changes minimal and isolated I created a new L-function class Lfunction_HMFDB that loads the L-function from the database and will fail if it is not present -- this is called from the routing function l_function_hmf_page inside a try/except and will catch the error and ignore it (this is a small hack that will go away once all HMF functions are stored in the database.  The case of Artin L-functions is handled similarly.

To test, compare:

http://127.0.0.1:37777/L/ModularForm/GL2/TotallyReal/2.2.5.1/holomorphic/2.2.5.1-500.1-a/0/0/
http://beta.lmfdb.org/L/ModularForm/GL2/TotallyReal/2.2.5.1/holomorphic/2.2.5.1-500.1-a/0/0/

and also try clicking from the L-function page to the HMF and back, test the bread, click through to different origins and then back via the origin page L-function links etc. (other than the URL the L-function page should look the same now matter which origin you came from).  In particular, the L-function page should list all the related objects now matter how the page was reached.

Note that this change does not only impact HMFs whose L-functions arise as genus 2 L-functions that are in the database, it also catches L-functions of elliptic curves over number fields, e.g. compare

http://127.0.0.1:37777/L/ModularForm/GL2/TotallyReal/2.2.5.1/holomorphic/2.2.5.1-31.1-a/0/0/
http://beta.lmfdb.org/L/ModularForm/GL2/TotallyReal/2.2.5.1/holomorphic/2.2.5.1-31.1-a/0/0/

For an Artin L-function example, compare

http://127.0.0.1:37777/L/ArtinRepresentation/2.23.3t2.1c1/
http://beta.lmfdb.org/L/ArtinRepresentation/2.23.3t2.1c1/

(again clicking through to the Artin rep and back should work).  Note that this is currently the only Artin L-function for which I have added an instance record, but I will be adding more later today (this doesn't involve any code changes so you can test the PR now).

Examples of L-function pages that should not change (i.e. they should still be rendered on the fly) because we have not precomputed them via another origin:

http://127.0.0.1:37777/L/ModularForm/GL2/TotallyReal/2.2.5.1/holomorphic/2.2.5.1-101.2-a/0/0/
http://beta.lmfdb.org/L/ModularForm/GL2/TotallyReal/2.2.5.1/holomorphic/2.2.5.1-101.2-a/0/0/

http://127.0.0.1:37777/L/ArtinRepresentation/2.2e8_3e2.8t5.2c1
http://beta.lmfdb.org/L/ArtinRepresentation/3.229.4t5.1c1/
